### PR TITLE
coreboot-utils: init at 4.10

### PR DESCRIPTION
**Pull request by ghost**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by petabyteboy** [comment]
@worldofpeace 
I am aware it is not best practise to use custom phases, but I think the advantage here is that new tools can be easily added. If you see any better way to do this, I would like to see it.

**Comment by petabyteboy** [comment]
One alternative I see is putting each tool in an own derivation, but using a generic expression for them. Do you think that's preferable?

File: pkgs/tools/misc/corebootutils/default.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
+{ stdenv, fetchurl, zlib, pciutils, coreutils, pmtools, iasl }:
+
+stdenv.mkDerivation rec {
+  pname = "corebootutils";
```
Comment (matthewbauer) [inline pkgs/tools/misc/corebootutils/default.nix]: Most distros are calling this "coreboot-utils":

https://repology.org/project/coreboot-utils/versions

Comment (petabyteboy) [reply pkgs/tools/misc/corebootutils/default.nix]: changed, thanks

**Comment by petabyteboy** [comment]
I don't get why ofborg is failing

**Comment by petabyteboy** [comment]
Ah, pmtools is an alias.

**Comment by worldofpeace** [comment]
> @worldofpeace
> I am aware it is not best practice to use custom phases, but I think the advantage here is that new tools can be easily added. If you see any better way to do this, I would like to see it.

I'd say ideally, coreboot would just have a `Makefile` in the `util` directory with targets for each of the utilities.


> One alternative I see is putting each tool in an own derivation, but using a generic expression for them. Do you think that's preferable?

I'm not sure about putting each tool into it's own derivation. 
Current disadvantage of `coreboot-utils` is that you can't have just one utility, so maybe split it.
Though it does seem to be more popular 
* https://repology.org/project/coreboot-utils/versions

My only nit currently is that it looks a bit procedural, we could use loops to make it clearer.
Fake code looks like
```
makeFlags = [ 
  "PREFIX=${placeholder "out"}"
];

preBuild = ''
  cd util
'';

buildPhase = ''
  runHook preBuild

  for util in ...directories of tools; do
    make $makeFlags -C $util
  done

  runHook postBuild
'';

installPhase = ''
  ....similar thing as build phase
'';

postFixup = ''
  ...should wrap that program here
'';

**Comment by petabyteboy** [comment]
@worldofpeace: I found it easier to solve it like this, please take a look

**Review by matthewbauer — APPROVED** [review]

File: pkgs/tools/misc/coreboot-utils/default.nix [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    inherit pname version meta;
+
+    src = fetchurl {
+      url = "https://coreboot.org/releases/coreboot-${version}.tar.xz";
+      sha256 = "1jsiz17afi2lqg1jv6lsl8s05w7vr7iwgg86y2qp369hcz6kcwfa";
+    };
+
+    enableParallelBuilding = true;
+
+    prePatch = ''
```
Comment (worldofpeace) [inline pkgs/tools/misc/coreboot-utils/default.nix]: `preBuild`?

File: pkgs/tools/misc/coreboot-utils/default.nix [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      cd ${path}
+    '';
+
+    configureFlags = [
+      "INSTALL=install"
+      "PREFIX=${placeholder "out"}"
+    ];
+    installFlags = configureFlags;
```
Comment (worldofpeace) [inline pkgs/tools/misc/coreboot-utils/default.nix]: I don't think this is needed, since anything in `makeFlags` will already be there.

File: pkgs/tools/misc/coreboot-utils/default.nix [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    makeFlags = configureFlags;
+  } // args);
+
+  utils = {
+    msrtool = generic {
+      pname = "msrtool";
+      meta.description = "Dump chipset-specific MSR registers";
+      buildInputs = [ pciutils zlib ];
+      preConfigure = "export INSTALL=install";
```
Comment (worldofpeace) [inline pkgs/tools/misc/coreboot-utils/default.nix]: Why was this needed?  (I think the configure script for this one wanted this)

File: pkgs/tools/misc/coreboot-utils/default.nix [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    amdfwtool = generic {
+      pname = "amdfwtool";
+      meta.description = "Create AMD firmware combination";
+      installPhase = "install -Dm755 amdfwtool $out/bin/amdfwtool";
+    };
+    acpidump-all = generic {
+      pname = "acpidump-all";
+      path = "util/acpi";
+      meta.description = "Walk through all ACPI tables with their addresses";
+      buildInputs = [ makeWrapper ];
```
Comment (worldofpeace) [inline pkgs/tools/misc/coreboot-utils/default.nix]:
```suggestion
      nativeBuildInputs = [ makeWrapper ];
```

File: pkgs/tools/misc/coreboot-utils/default.nix [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      buildPhase = " ";
+      installPhase = "install -Dm755 acpidump-all $out/bin/acpidump-all";
+      postFixup = ''wrapProgram "$out/bin/acpidump-all" --set PATH "${stdenv.lib.makeBinPath [coreutils acpica-tools iasl gnugrep gnused file]}"'';
```
Comment (worldofpeace) [inline pkgs/tools/misc/coreboot-utils/default.nix]: Can you actually make this multline, or use `"` and escape with `\`

this might look nice
```
postFixup = let 
  binPath = [ coreutils  acpica-tools iasl gnugrep  gnused  file ];
in "wrapProgram $out/bin/acpidump-all --set PATH ${stdenv.lib.makeBinPath binPath}";

File: pkgs/tools/misc/coreboot-utils/default.nix (hunk 3 above)
Comment (petabyteboy) [reply pkgs/tools/misc/coreboot-utils/default.nix]: Yup you're right!

File: pkgs/tools/misc/coreboot-utils/default.nix (hunk 4 above)
Comment (petabyteboy) [reply pkgs/tools/misc/coreboot-utils/default.nix]: Yes, it ignores it when it comes as a parameter, but uses it when it's in the environment.

File: pkgs/tools/misc/coreboot-utils/default.nix (hunk 5 above)
Comment (petabyteboy) [reply pkgs/tools/misc/coreboot-utils/default.nix]: Thanks

File: pkgs/tools/misc/coreboot-utils/default.nix (hunk 2 above)
Comment (petabyteboy) [reply pkgs/tools/misc/coreboot-utils/default.nix]: Using preConfigure there will make it hard to use preConfigure in any of the specific derivations, like msrtool. I will change it to postPatch instead, so that patches for the coreboot-tree can be used.

File: pkgs/tools/misc/coreboot-utils/default.nix (hunk 6 above)
Comment (petabyteboy) [reply pkgs/tools/misc/coreboot-utils/default.nix]: Thanks

Comment (petabyteboy) [reply pkgs/tools/misc/coreboot-utils/default.nix]: Thanks

**Comment by petabyteboy** [comment]
We should be good to go then :)

**Comment by worldofpeace** [comment]
I noticed when building i had to unpack the archive multiple times, but I guess there's no way around that.

File: pkgs/tools/misc/coreboot-utils/default.nix [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+      pname = "amdfwtool";
+      meta.description = "Create AMD firmware combination";
+      installPhase = "install -Dm755 amdfwtool $out/bin/amdfwtool";
+    };
+    acpidump-all = generic {
+      pname = "acpidump-all";
+      path = "util/acpi";
+      meta.description = "Walk through all ACPI tables with their addresses";
+      nativeBuildInputs = [ makeWrapper ];
+      buildPhase = " ";
```
Comment (worldofpeace) [inline pkgs/tools/misc/coreboot-utils/default.nix]:
```suggestion
      dontBuild = true;
```

Comment (petabyteboy) [reply pkgs/tools/misc/coreboot-utils/default.nix]: Done

**Comment by worldofpeace** [comment]
It looks like it tries to install things to `sbin`
```
├── bin
│   ├── acpidump-all -> /nix/store/50ydys5lgnqbv34rvpra2jwa7ys85g3l-acpidump-all-4.10/bin/acpidump-all
│   ├── amdfwtool -> /nix/store/b78wwi966a4an64gk19nhmkblzqp42wv-amdfwtool-4.10/bin/amdfwtool
│   ├── cbfs-compression-tool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/cbfs-compression-tool
│   ├── cbfstool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/cbfstool
│   ├── cbmem -> /nix/store/3nbj9bmrdnvg9s3a919kq3ns2wil93lz-cbmem-4.10/bin/cbmem
│   ├── ectool -> /nix/store/ypinf2l28mn05j5jdzx5szx5i0fx8dly-ectool-4.10/bin/ectool
│   ├── fmaptool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/fmaptool
│   ├── ifdtool -> /nix/store/skwh2bdgjppksr4sn5nmw0cp4h4qnpxk-ifdtool-4.10/bin/ifdtool
│   ├── ifittool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/ifittool
│   ├── ifwitool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/ifwitool
│   ├── intelmetool -> /nix/store/hsf1rgq5aqv2bhjhlka1n574wzjaa5i2-intelmetool-4.10/bin/intelmetool
│   ├── inteltool -> /nix/store/8m36y393g2ikfnk9y04h84kklvif8jyw-inteltool-4.10/bin/inteltool
│   ├── msrtool -> /nix/store/cayfmg47wgliwmwz5cn77hmvwphr2601-msrtool-4.10/bin/msrtool
│   ├── nvramtool -> /nix/store/kcysznlapfr4cj3vz8nd6k8hbhaz3pgv-nvramtool-4.10/bin/nvramtool
│   ├── rmodtool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/rmodtool
│   └── superiotool -> /nix/store/l28ay2p1ly63nz9dw3bddj8ls0z4r85d-superiotool-4.10/bin/superiotool
├── sbin
│   ├── cbmem -> /nix/store/3nbj9bmrdnvg9s3a919kq3ns2wil93lz-cbmem-4.10/sbin/cbmem
│   ├── ectool -> /nix/store/ypinf2l28mn05j5jdzx5szx5i0fx8dly-ectool-4.10/sbin/ectool
│   ├── intelmetool -> /nix/store/hsf1rgq5aqv2bhjhlka1n574wzjaa5i2-intelmetool-4.10/sbin/intelmetool
│   ├── inteltool -> /nix/store/8m36y393g2ikfnk9y04h84kklvif8jyw-inteltool-4.10/sbin/inteltool
│   ├── msrtool -> /nix/store/cayfmg47wgliwmwz5cn77hmvwphr2601-msrtool-4.10/sbin/msrtool
│   ├── nvramtool -> /nix/store/kcysznlapfr4cj3vz8nd6k8hbhaz3pgv-nvramtool-4.10/sbin/nvramtool
│   └── superiotool -> /nix/store/l28ay2p1ly63nz9dw3bddj8ls0z4r85d-superiotool-4.10/sbin/superiotool
```

and we have setup hook that moves these things, but it probably doesn't work because of duplication?
Perhaps look at which makefiles do this and install manually, or maybe there's a way to override it.

**Comment by petabyteboy** [comment]
> It looks like it tries to install things to `sbin`
> 
> ```
> ├── bin
> │   ├── acpidump-all -> /nix/store/50ydys5lgnqbv34rvpra2jwa7ys85g3l-acpidump-all-4.10/bin/acpidump-all
> │   ├── amdfwtool -> /nix/store/b78wwi966a4an64gk19nhmkblzqp42wv-amdfwtool-4.10/bin/amdfwtool
> │   ├── cbfs-compression-tool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/cbfs-compression-tool
> │   ├── cbfstool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/cbfstool
> │   ├── cbmem -> /nix/store/3nbj9bmrdnvg9s3a919kq3ns2wil93lz-cbmem-4.10/bin/cbmem
> │   ├── ectool -> /nix/store/ypinf2l28mn05j5jdzx5szx5i0fx8dly-ectool-4.10/bin/ectool
> │   ├── fmaptool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/fmaptool
> │   ├── ifdtool -> /nix/store/skwh2bdgjppksr4sn5nmw0cp4h4qnpxk-ifdtool-4.10/bin/ifdtool
> │   ├── ifittool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/ifittool
> │   ├── ifwitool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/ifwitool
> │   ├── intelmetool -> /nix/store/hsf1rgq5aqv2bhjhlka1n574wzjaa5i2-intelmetool-4.10/bin/intelmetool
> │   ├── inteltool -> /nix/store/8m36y393g2ikfnk9y04h84kklvif8jyw-inteltool-4.10/bin/inteltool
> │   ├── msrtool -> /nix/store/cayfmg47wgliwmwz5cn77hmvwphr2601-msrtool-4.10/bin/msrtool
> │   ├── nvramtool -> /nix/store/kcysznlapfr4cj3vz8nd6k8hbhaz3pgv-nvramtool-4.10/bin/nvramtool
> │   ├── rmodtool -> /nix/store/f1qvacrckyvnxshkdpl37gzf06zlhl2s-cbfstool-4.10/bin/rmodtool
> │   └── superiotool -> /nix/store/l28ay2p1ly63nz9dw3bddj8ls0z4r85d-superiotool-4.10/bin/superiotool
> ├── sbin
> │   ├── cbmem -> /nix/store/3nbj9bmrdnvg9s3a919kq3ns2wil93lz-cbmem-4.10/sbin/cbmem
> │   ├── ectool -> /nix/store/ypinf2l28mn05j5jdzx5szx5i0fx8dly-ectool-4.10/sbin/ectool
> │   ├── intelmetool -> /nix/store/hsf1rgq5aqv2bhjhlka1n574wzjaa5i2-intelmetool-4.10/sbin/intelmetool
> │   ├── inteltool -> /nix/store/8m36y393g2ikfnk9y04h84kklvif8jyw-inteltool-4.10/sbin/inteltool
> │   ├── msrtool -> /nix/store/cayfmg47wgliwmwz5cn77hmvwphr2601-msrtool-4.10/sbin/msrtool
> │   ├── nvramtool -> /nix/store/kcysznlapfr4cj3vz8nd6k8hbhaz3pgv-nvramtool-4.10/sbin/nvramtool
> │   └── superiotool -> /nix/store/l28ay2p1ly63nz9dw3bddj8ls0z4r85d-superiotool-4.10/sbin/superiotool
> ```
> 
> and we have setup hook that moves these things, but it probably doesn't work because of duplication?
> Perhaps look at which makefiles do this and install manually, or maybe there's a way to override it.

It moves the files, and then creates a symlink from sbin to bin so that things don't break. Is this a problem?

**Comment by worldofpeace** [comment]
> It moves the files, and then creates a symlink from sbin to bin so that things don't break. Is this a problem?

Ahh, I forgot it was a `buildEnv`. So what happens is it works properly individually, but when they get merged with `buildEnv` `sbin` becomes an actual directory with symlinks to symlinks.
I think that's kinda messy for `coreboot-utils`, can you prune that away?

**Comment by petabyteboy** [comment]
> > It moves the files, and then creates a symlink from sbin to bin so that things don't break. Is this a problem?
> 
> Ahh, I forgot it was a `buildEnv`. So what happens is it works properly individually, but when they get merged with `buildEnv` `sbin` becomes an actual directory with symlinks to symlinks.
> I think that's kinda messy for `coreboot-utils`, can you prune that away?

Sure, done.

**Review by worldofpeace — APPROVED** [review]
LGTM, can you describe your changes in your commit body?

**Comment by petabyteboy** [comment]
> LGTM, can you describe your changes in your commit body?

Done

**Comment by worldofpeace** [comment]
Thanks for fixing this @petabyteboy  :sparkles:
